REGRESSION(299621@main): [MacOS Debug] ASSERTION FAILED: m_gamepads[platformGamepad.index] in 'gamepad/gamepad-visibility-1.html' is failing

kkinnunen-apple opened the pull request:
#### 1b7f230c1a6e01d76f98af8b9e0dc5f077e8da3f
<pre>
REGRESSION(<a href="https://commits.webkit.org/299621@main">299621@main</a>): [MacOS Debug] ASSERTION FAILED: m_gamepads[platformGamepad.index] in &apos;gamepad/gamepad-visibility-1.html&apos; is failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=298476">https://bugs.webkit.org/show_bug.cgi?id=298476</a>
<a href="https://rdar.apple.com/159970513">rdar://159970513</a>

Reviewed by NOBODY (OOPS!).

Test runs such as following would fail:
run-webkit-tests --iterations 100 --no-build --verbose  --details \
 --debug-rwt-logging --no-retry --no-show-results --debug \
 --exit-after-n-failures=1 gamepad/gamepad-vibrationActuator-type.html \
 gamepad/gamepad-visibility-1.html

Running gamepad/gamepad-visibility-1.html would result frequently in
crash of the subsequent test, making it harder to analyze test failures.

GamepadManager would try to implement the feature where gamepad
connection info is not exposed to the page unless the gamepad is
interacted with. It would assert when running tests, loading two pages
at the same time.

The first web view would see mock gamepad connect, use, disconnect.
Then other web view would invoke connect, disconnect.
The other web view would survive the disconnect, as it would be stored
in &quot;blind&quot; list. The first web view would be moved off of the &quot;blind
list&quot;, and would assert on seeing disconnect without connect. Connects
are no-op due to the above exposure filtering.

Fix by removing complex list handling related to the &quot;blind&quot;
window/navigator concept. Also remove list handling related to vision
os &quot;guarantine&quot; feature.

Instead, implement [[exposed]] property as designed in the spec:
When it&apos;s time to expose the Gamepad, create it to NavigatorGamepad.

To be able to implement consistent code, fix existing bugs:
 - Disconnected Gamepad instances were left in &quot;connected&quot; state. Mark
   them disconnected before dispatching the disconnect event.
 - Navigator.getGamepads() list would end in null values. It
   is specified to end in a populated value or be empty.
 - The disconnected gamepad should remain in Navigator.getGamepads()
   for the duration of the gamepaddisconnected event.

Tests: gamepad/gamepad-connect-disconnect-invisible.html
       gamepad/gamepad-visibility-poll.html

Tests: gamepad/gamepad-connect-disconnect-invisible.html
       gamepad/gamepad-visibility-poll.html
* LayoutTests/gamepad/gamepad-connect-disconnect-invisible-expected.txt: Added.
* LayoutTests/gamepad/gamepad-connect-disconnect-invisible.html: Added.
* LayoutTests/gamepad/gamepad-polling-access-expected.txt:
* LayoutTests/gamepad/gamepad-visibility-poll-expected.txt: Added.
* LayoutTests/gamepad/gamepad-visibility-poll.html: Added.
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/Modules/gamepad/GamepadManager.cpp:
(WebCore::GamepadManager::GamepadManager):
(WebCore::GamepadManager::platformGamepadConnected):
(WebCore::GamepadManager::platformGamepadDisconnected):
(WebCore::GamepadManager::platformGamepadInputActivity):
(WebCore::GamepadManager::makeGamepadsVisible):
(WebCore::GamepadManager::registerNavigator):
(WebCore::GamepadManager::unregisterNavigator):
(WebCore::GamepadManager::registerDOMWindow):
(WebCore::GamepadManager::unregisterDOMWindow):
(WebCore::GamepadManager::didUpdateGamepadAccess):
(WebCore::navigatorGamepadFromDOMWindow): Deleted.
(WebCore::GamepadManager::findUnquarantinedNavigatorsAndWindows): Deleted.
(WebCore::GamepadManager::makeGamepadVisible): Deleted.
(WebCore::GamepadManager::updateQuarantineStatus): Deleted.
* Source/WebCore/Modules/gamepad/GamepadManager.h:
* Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
(WebCore::NavigatorGamepad::ensureGamepad):
(WebCore::NavigatorGamepad::gamepadIfExists const):
(WebCore::NavigatorGamepad::removeGamepadIfExists):
(WebCore::NavigatorGamepad::gamepads):
(WebCore::NavigatorGamepad::gamepadFromPlatformGamepad):
(WebCore::NavigatorGamepad::gamepadsBecameVisible): Deleted.
(WebCore::NavigatorGamepad::gamepadConnected): Deleted.
(WebCore::NavigatorGamepad::gamepadDisconnected): Deleted.
(WebCore::NavigatorGamepad::protectedPage const): Deleted.
* Source/WebCore/Modules/gamepad/NavigatorGamepad.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::allowGamepadAccess):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.cpp:
(WebKit::RemoteMediaPlayerManager::setUseGPUProcess):
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.cpp:
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::getVideoFrameBuffer):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b7f230c1a6e01d76f98af8b9e0dc5f077e8da3f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6707 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45409 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141776 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86257 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/cedb727f-e34e-408c-8bda-1545a752fe47) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136066 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7325 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6571 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102616 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69928 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137143 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5104 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120277 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83411 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1de07d68-1f76-4331-96e4-336e20cb23b5) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4980 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2568 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/1592 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114171 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38413 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144423 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6377 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38991 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111000 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6460 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5362 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111242 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4790 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116550 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60148 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6429 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34767 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6277 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69895 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6518 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6383 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/bfbaebe34f89412f9943a012eddbed53b51e6c0b)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133790 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-26-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135660 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136737 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | | ⏳ 🛠 vision-apple 
| | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac-debug ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144013 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-26-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/144013 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/144013 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-26-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-26-Build-EWS "Waiting in queue, processing has not started yet") | | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | | </details>

beidson commented:
Removing the blind list is a huge red flag to me - Regardless of the spec here, we explicitly intend to be more private.
Even if you get another review here, please hold off on landing until I have a chance to read up on what "Exposed" means so I can develop an opinion on this.

kkinnunen-apple commented:
> Removing the blind list is a huge red flag to me - Regardless of the spec here, we explicitly intend to be more private. Even if you get another review here, please hold off on landing until I have a chance to read up on what "Exposed" means so I can develop an opinion on this.

Thanks. You don't need to get hung up on [[exposed]]. Those are more spec vehicles so that the reader understands what is the semantics of the spec.

I simplified the patch a bit so that the [[exposed]] does not block.

kkinnunen-apple committed:
REGRESSION(299621@main): [MacOS Debug] ASSERTION FAILED: m_gamepads[platformGamepad.index] in 'gamepad/gamepad-visibility-1.html' is failing
https://bugs.webkit.org/show_bug.cgi?id=298476
rdar://159970513

Reviewed by NOBODY (OOPS!).

Test runs such as following would fail:
run-webkit-tests --iterations 100 --no-build --verbose  --details \
 --debug-rwt-logging --no-retry --no-show-results --debug \
 --exit-after-n-failures=1 gamepad/gamepad-vibrationActuator-type.html \
 gamepad/gamepad-visibility-1.html

Running gamepad/gamepad-visibility-1.html would result frequently in
crash of the subsequent test, making it harder to analyze test failures.

GamepadManager would try to implement the feature where gamepad
connection info is not exposed to the page unless the gamepad is
interacted with. It would assert when running tests, loading two pages
at the same time.

The first web view would see mock gamepad connect, use, disconnect.
Then other web view would invoke connect, disconnect.
The other web view would survive the disconnect, as it would be stored
in "blind" list. The first web view would be moved off of the "blind
list", and would assert on seeing disconnect without connect. Connects
are no-op due to the above exposure filtering.

Fix by removing complex list handling related to the "blind"
window/navigator concept. Also remove list handling related to vision
os "guarantine" feature.

Instead, implement [[exposed]] property as designed in the spec:
When it's time to expose the Gamepad, create it to NavigatorGamepad.

To be able to implement consistent code, fix existing bugs:
 - Disconnected Gamepad instances were left in "connected" state. Mark
   them disconnected before dispatching the disconnect event.
 - Navigator.getGamepads() list would end in null values. It
   is specified to end in a populated value or be empty.
 - The disconnected gamepad should remain in Navigator.getGamepads()
   for the duration of the gamepaddisconnected event.

Tests: gamepad/gamepad-connect-disconnect-invisible.html
       gamepad/gamepad-visibility-poll.html

Tests: gamepad/gamepad-connect-disconnect-invisible.html
       gamepad/gamepad-visibility-poll.html
* LayoutTests/gamepad/gamepad-connect-disconnect-invisible-expected.txt: Added.
* LayoutTests/gamepad/gamepad-connect-disconnect-invisible.html: Added.
* LayoutTests/gamepad/gamepad-polling-access-expected.txt:
* LayoutTests/gamepad/gamepad-visibility-poll-expected.txt: Added.
* LayoutTests/gamepad/gamepad-visibility-poll.html: Added.
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/Modules/gamepad/GamepadManager.cpp:
(WebCore::GamepadManager::GamepadManager):
(WebCore::GamepadManager::platformGamepadConnected):
(WebCore::GamepadManager::platformGamepadDisconnected):
(WebCore::GamepadManager::platformGamepadInputActivity):
(WebCore::GamepadManager::makeGamepadsVisible):
(WebCore::GamepadManager::registerNavigator):
(WebCore::GamepadManager::unregisterNavigator):
(WebCore::GamepadManager::registerDOMWindow):
(WebCore::GamepadManager::unregisterDOMWindow):
(WebCore::GamepadManager::didUpdateGamepadAccess):
(WebCore::navigatorGamepadFromDOMWindow): Deleted.
(WebCore::GamepadManager::findUnquarantinedNavigatorsAndWindows): Deleted.
(WebCore::GamepadManager::makeGamepadVisible): Deleted.
(WebCore::GamepadManager::updateQuarantineStatus): Deleted.
* Source/WebCore/Modules/gamepad/GamepadManager.h:
* Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
(WebCore::NavigatorGamepad::ensureGamepad):
(WebCore::NavigatorGamepad::gamepadIfExists const):
(WebCore::NavigatorGamepad::removeGamepadIfExists):
(WebCore::NavigatorGamepad::gamepads):
(WebCore::NavigatorGamepad::gamepadFromPlatformGamepad):
(WebCore::NavigatorGamepad::gamepadsBecameVisible): Deleted.
(WebCore::NavigatorGamepad::gamepadConnected): Deleted.
(WebCore::NavigatorGamepad::gamepadDisconnected): Deleted.
(WebCore::NavigatorGamepad::protectedPage const): Deleted.
* Source/WebCore/Modules/gamepad/NavigatorGamepad.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::allowGamepadAccess):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.cpp:
(WebKit::RemoteMediaPlayerManager::setUseGPUProcess):
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.cpp:
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::getVideoFrameBuffer):
